### PR TITLE
[feature] delayed_sidekiq strategy

### PR DESCRIPTION
**Pull request by skcc321**
# What does it do?
I added a new strategy - `delayed_sidekiq`.
It behaves the next way:
- the strategy has a `latency` option (for instance 10 seconds), `margin` & `reindex_wrapper`
- when you mutate active record objects very frequently the strategy will accumulate ids of mutated records during the latency window.
- that means that we are going to reindex all changes that happened during 10 seconds at once.
- such a latency gives us the opportunity to use "read replica" instead of "primary" DB (RoR multi DB connection)
- as a result, if you don't need strong consistency you can define an appropriate latency window and gain significant performance improvement (fewer calls to DB & replica usage)
- `margin` option is important to cover `replication_lag`

an example:
```ruby
class CitiesIndex < Chewy::Index
  index_scope City
  
  strategy_config delayed_sidekiq: {
    latency: 10,  
    margin: 2,
    reindex_wrapper: ->(&reindex) { ActiveRecord::Base.connected_to(role: :reading) { reindex.call }} # fetch data from read replica instead of the primary node
  }
  
  field :name
  field :address
  ...
end

class City < ActiveRecord::Model
  update_index("cities") { self if persistently_changed? }
  ...
end

------ Request 1 -------
Chewy.strategy(:delayed_sidekiq) do
  City.create(name: "Amsterdam") # created city with id=1
end
------ Request 2 -------
Chewy.strategy(:delayed_sidekiq) do
  City.create(name: "Paris") # created city with id=2
end
------ Request 3 -------
Chewy.strategy(:delayed_sidekiq) do
  City.create(name: "Rome") # created city with id=3
end
```

those three parallel requests that happened during the "latency" window are going to schedule a single worker to call import with all accumulated ids 1, 2, 3
```ruby
CityIndex.import([1, 2, 3]) # sql request will go to { role: :reading } due to definition in the index
```

## Strategy option in #import method.

```ruby
CityIndex.import([1, 2, 3], strategy: :delayed_sidekiq)
```

## Accumulation of :update_fields option between chunks.
```ruby
CityIndex.import([1, 2], update_fields: ["name"], strategy: :delayed_sidekiq) 
CityIndex.import([3], update_fields: ["address"], strategy: :delayed_sidekiq)
```
will cause
```ruby
CityIndex.import([1, 2, 3], update_fields: ["name", "address"])
```
that can be very useful if you have some complex crutches in the index definition and you don't want to load the database with obviously useless SQL queries.

## Configuration
- at `CityIndex` level
```ruby
class CitiesIndex < Chewy::Index
  index_scope City
  
 strategy_config delayed_sidekiq: {
    latency: 10,  
    margin: 2,
    reindex_wrapper: ->(&reindex) { ActiveRecord::Base.connected_to(role: :reading) { reindex.call }} # fetch data from read replica instead of the primary node
 }
```
- at the initializer level
```ruby
Chewy.settings = {
  strategy_config: {
    delayed_sidekiq: {
      reindex_wrapper: ->(&import) { ActiveRecord::Base.connected_to(role: :reading) { import.call } },
      margin: 2, # replication lag seconds
      latency: 10 # accumulation window seconds
    }
  }
}
```
- at chewy.yml level
```yaml
strategy_config:
  delayed_sidekiq:
     latency: 10
     margin: 10
     # reindex_wrapper option is not supported in YAML config. Use either initializer either index file instead.
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the changelog if the new code introduces user-observable changes. See [changelog entry format](https://github.com/toptal/chewy/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by skcc321** [comment]
@AlexVPopov could you please review this one please? (or point someone who can take a look)

**Comment by skcc321** [comment]
@konalegi one more time, please. Got failed CI due to no Redis service running.

**Comment by konalegi** [comment]
@skcc321 Thank you very much for the PR,  I'll try to review this in a week or find someone who can do it.

**Comment by skcc321** [comment]
@konalegi one more time CI, please (I hope the last one) - fixed linters & ruby 2.x related failed tests.

**Comment by skcc321** [comment]
@konalegi sorry for bothering you. Any updates on that one?

**Comment by konalegi** [comment]
@skcc321, sorry, everyone is busy these days. I'll try to review this on Thursday/Friday, added my Todo list

File: README.md
Comment (mrzasa) [inline README.md:779]:
```suggestion
It accumulates ids of records to be reindexed during the latency window in redis and then does the reindexing of all accumulated records at once.
```

File: README.md
Comment (mrzasa) [inline README.md:781]:
```suggestion
It supports `update_fields` option, so it will try to select just enough data from the DB
```

File: README.md
Comment (mrzasa) [inline README.md:806]:
```suggestion
Explicit call of the reindex using `:delayed_sidekiq strategy`
```

File: README.md
Comment (mrzasa) [inline README.md:811]:
```suggestion
Explicit call of the reindex using `:delayed_sidekiq` strategy with `:update_fields` support
```

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (mrzasa) [inline lib/chewy/strategy/delayed_sidekiq.rb:45]: Why do you prefer to monkeypath when you can edit the files directly?

File: spec/chewy/strategy/delayed_sidekiq_spec.rb
Comment (mrzasa) [inline spec/chewy/strategy/delayed_sidekiq_spec.rb:70]: Could you add meaningful names to the test cases? I know we have  anonymous test in the chewy test suite, but it'd be better to avoid them unless the case is obvious.

**Review by mrzasa — COMMENTED** [review]
Thanks for the contribution. I've added some early comments, I didn't have time to test it properly.

File: spec/chewy/strategy/delayed_sidekiq_spec.rb
Comment (skcc321) [reply spec/chewy/strategy/delayed_sidekiq_spec.rb:70]: makes sense. I'll do!

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb:45]: No problem. I can easily add these changes to Chewy::Index if you are OK with that.
I copied that stuff from the `patch file` I have in the project so that is why everything exists in the single file.

**Comment by skcc321** [comment]
@mrzasa thank you for the review. I pushed the update. Could you take a look again, please?

**Comment by skcc321** [comment]
@mrzasa thank you for the review. I pushed all suggested changes in a separate commit (just for easier review next time).
Could you take a look one more time and restart the CI, please?

**Comment by skcc321** [comment]
@konalegi sorry for bothering you, could you please trigger the CI again?

**Review by konalegi — COMMENTED** [review]
Sorry for making you wait for so long.
Looks good to me, I'll need to do some manual testing, if everything will be nice, I'll merge it.

**Comment by skcc321** [comment]
Thank you @konalegi. I'll work on your comments soon.

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb:103]: Yep. I don't want to schedule tons of jobs. I want to have a single job during the specific time window based on the `latency` value.
Let's take a look at this example:
```ruby
def simulate(latency) # `at` code wrapped by dummy `simulate` method.
     schedule_at = latency.seconds.from_now.to_f
     (schedule_at - (schedule_at % latency)).to_i
end

latency = 2 # in seconds

# simulate sequential execution of the method
10.times { |i| puts "current_time: #{Time.current.to_i}, at: #{simulate(latency)}"; sleep(0.7) }                                                                                                  
current_time: 1679347458, at: 1679347460
current_time: 1679347459, at: 1679347460
current_time: 1679347459, at: 1679347460
current_time: 1679347460, at: 1679347462
current_time: 1679347461, at: 1679347462
current_time: 1679347461, at: 1679347462
current_time: 1679347462, at: 1679347464
current_time: 1679347463, at: 1679347464
current_time: 1679347463, at: 1679347464
current_time: 1679347464, at: 1679347466
```
now imagine that `reindex` is triggered at `current_time`. I do not schedule reindex at current_time, instead of that I accumulate all `ids` to reindex during the latency time window and schedule the reindex job at `at` value.
This is a very simple trick that is a core of the `delayed_sidekiq` strategy.

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb]: done.

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb]: added a comment near the method name.

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb]: done

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (skcc321) [reply lib/chewy/strategy/delayed_sidekiq.rb]: done

File: lib/chewy/strategy/delayed_sidekiq.rb
Comment (konalegi) [reply lib/chewy/strategy/delayed_sidekiq.rb]: makes sense, missed, thanks for clarification

**Comment by konalegi** [comment]
LGTM, please address my last comment and rubocop offense.
Thanks for great feature!

File: README.md
Comment (konalegi) [inline README.md:795]: sorry, this one as well please :)

Comment (skcc321) [reply README.md]: Oh, damn. Thank you! Missed that one.
Now it should be fine.

**Review by konalegi — APPROVED** [review]

**Comment by skcc321** [comment]
@mrzasa could approve please (if you have no concerns) :)

**Comment by konalegi** [comment]
Since this PR doesn't modify any existing functionality and brings new feature going to be merge right away.

**Comment by konalegi** [comment]
Released

**Pull request merged**
